src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov.hpp

```diff
@@ -76,101 +66,127 @@ struct Esirkepov<T_ParticleShape, DIM3>
-                                      float_X(math::floor(pos.z() * float_X(2.0)))
-                                      );
-            cursorJ = cursorJ(
-                              PMacc::math::Int < 3 > (
-                                                      coordinate_shift.x(),
-                                                      coordinate_shift.y(),
-                                                      coordinate_shift.z()
-                                                      ));
-            //same as: pos = pos - coordinate_shift;
-            line.m_pos0 -= (coordinate_shift);
-            line.m_pos1 -= (coordinate_shift);
+            int iStart;
+            int iEnd;
+            constexpr bool isSupportEven = ( supp % 2 == 0 );
+            RelayPoint< isSupportEven >()(
+                iStart,
+                iEnd,
+                line.m_pos0[d],
+                line.m_pos1[d]
+            );
+            gridShift[d] = iStart < iEnd ? iStart : iEnd; // integer min function
+            /* particle is leaving the cell */
+            leaveCell[d] = iStart != iEnd ? 1 : 0;
+            /* shift the particle position to the virtual coordinate system */
+            line.m_pos0[d] -= gridShift[d];
+            line.m_pos1[d] -= gridShift[d];
         }
-
+        /* shift current field to the virtual coordinate system */
+        auto cursorJ = dataBoxJ.shift(gridShift).toCursor();
         /**
          * \brief the following three calls separate the 3D current deposition
          * into three independent 1D calls, each for one direction and current component.
          * Therefore the coordinate system has to be rotated so that the z-direction
          * is always specific.
          */
         using namespace cursor::tools;
-        cptCurrent1D(twistVectorFieldAxes<PMacc::math::CT::Int < 1, 2, 0 > >(cursorJ), rotateOrigin < 1, 2, 0 > (line), cellSize.x());
-        cptCurrent1D(twistVectorFieldAxes<PMacc::math::CT::Int < 2, 0, 1 > >(cursorJ), rotateOrigin < 2, 0, 1 > (line), cellSize.y());
-        cptCurrent1D(cursorJ, line, cellSize.z());
+        cptCurrent1D(
+            DataSpace<simDim>(leaveCell.y(),leaveCell.z(),leaveCell.x()),
+            twistVectorFieldAxes<PMacc::math::CT::Int < 1, 2, 0 > >(cursorJ),
+            rotateOrigin < 1, 2, 0 > (line),
+            cellSize.x()
+        );
+        cptCurrent1D(
+            DataSpace<simDim>(leaveCell.z(),leaveCell.x(),leaveCell.y()),
+            twistVectorFieldAxes<PMacc::math::CT::Int < 2, 0, 1 > >(cursorJ),
+            rotateOrigin < 2, 0, 1 > (line),
+            cellSize.y()
+        );
+        cptCurrent1D(
+            leaveCell,
+            cursorJ,
+            line,
+            cellSize.z()
+        );
     }
 
     /**
      * deposites current in z-direction
+     * \param leaveCell vector with information (for each direction) if the particle is leaving the cell
```

Comment:
add: *\param leaveCell vector with information if the particle is leaving the cell (for each direction, __0 means stays in cell and 1 means leaves cell__)*